src/main/java/org/candlepin/model/ConsumerCurator.java

```diff
@@ -1786,4 +1791,61 @@ public int unlinkCaCertificates(Collection<String> certIds) {
 
         return updated;
     }
+
+    /**
+     * Takes a list of consumers and of each, updates its owner to the given one.
+     *
+     * @param consumerIds
+     *  consumer to be updated
+     * @param newOwner
+     *  owner to update the consumers to
+     */
+    @Transactional
+    public void bulkUpdateOwner(Collection<String> consumerIds, Owner newOwner) {
+        if (consumerIds == null || consumerIds.isEmpty()) {
+            return;
+        }
+        if (newOwner == null) {
+            throw new IllegalArgumentException("New owner is required!");
+        }
+
+        String query = """
+            UPDATE Consumer c \
+            SET c.owner.id = :ownerId \
+            WHERE c.id IN (:consumers)
+            """;
+
+        for (Collection<String> consumersBlock : this.partition(consumerIds)) {
+            this.currentSession().createQuery(query)
+                .setParameter("ownerId", newOwner.getId())
+                .setParameter("consumers", consumersBlock)
+                .executeUpdate();
+        }
+    }
+
+    public Collection<String> lockAndLoadIds(Collection<? extends Serializable> ids) {
+        if (ids == null || ids.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        // Sort and de-duplicate the provided collection of IDs so we have a deterministic locking
+        // order for the entities (helps avoid deadlock)
+        SortedSet<Serializable> idSet = new TreeSet<>(ids);
+
+        // Fetch the IDs of locked entities from the DB...
+        String hql = """
+            SELECT c.id FROM Consumer c
+            WHERE c.id IN (:ids)""";
+        Query<String> query = this.currentSession()
+            .createQuery(hql, String.class)
+            .setLockOptions(new LockOptions(LockMode.PESSIMISTIC_WRITE));
+
+        List<String> lockedIds = new ArrayList<>(idSet.size());
+        for (List<Serializable> idBlock : this.partition(idSet)) {
+            lockedIds.addAll(query.setParameter("ids", idBlock).getResultList());
+        }
+
+        return lockedIds;
+    }
```

Comment:
This seems to be working fine. Tested by running 20 jobs with a same set of consumers and locking for 10 seconds. Attempts to lock it with `byMultipleIds()` and entity graphs were unsuccessful. I couldn't figure out how to limit the amount of fetched data.

Reply:
It looks good to me, but I'd also ask @Ceiu who is more familiar with hibernate

Reply:
Not getting locks via `byMultipleIds` is very disconcerting to hear, since we have other locking calls and critical parts of the locking model that rely on it. But I'm also not entirely surprised to hear that it didn't let you modify the graph.

That aside, the only other ways of locking I know of via Hibernate is either this or a direct SQL query with `FOR UPDATE` (which the above code should be triggering).

Reply:
You can use these config properties to show the SQL generated: https://www.candlepinproject.org/docs/candlepin/logging.html#logging-sql-queries-and-parameters to see if the `FOR UPDATE` is there

Reply:
Yes. It is triggering FOR UPDATE. This is the generated sql.

```sql
select consumer0_.id as col_0_0_ from cp_consumer consumer0_ where consumer0_.id in (? , ?) for update
```